Jetpack Migration: Provide a path for JP to trigger data export in WP

dvdchr opened the pull request:
Refs qk5s4q4mZsXknwhMWo2vIY-fi-7315%3A20747, pcdRpT-1bA-p2#comment-2009

Designs:
<img width="565" alt="Screenshot 2022-12-03 at 01 59 00" src="https://user-images.githubusercontent.com/1299411/205366191-602f8626-ae07-4b6e-9541-e9576ea3e384.png">

This PR covers changes needed to support JP->WP->JP roundtrip for users that have compatible WordPress app but never triggered the export process once:

- Added a new `DataMigrationError`: `dataNotReadyToImport`. This can signal if we should check for a compatible WP app version and offer the user to export from WP if it exists.
- Added a new route, `WordPressExportRoute`. When this route is performed, it will start the export process and redirect back to JP upon completion. Note that this route should only be performed when the URL scheme is the one that's used to disable WP notifications.
- Added another new route, `JetpackImportRoute`. When this route is performed, it will attempt to restart the migration process from the Jetpack side and reconstruct the UI.
- Introduces `showLoadWordPressUI(schemeUrl:)` in `JetpackWindowManager` that presents a UI when the user should load WordPress to start the migration pre-flight.
- Adds a call to `MigratableStateTracker` when the import process starts

## To test

<details><summary><b>Prerequisites: Enable Feature Flags</b></summary>
<p>

**Via the app**
1. Clean install Jetpack and WordPress apps then log into your account
2. My Site > Me > App Settings > Debug
3. Enable "Content Migration" Feature Flag on both apps
4. Enable "Jetpack powered bottom sheet" Feature Flag on WordPress app

**Programmatically**
1. Go to `FeatureFlag.swift` file
2. Enable `contentMigration` and `jetpackPoweredBottomSheet` Feature Flags
3. Build and run both WordPress and Jetpack apps

</p>
</details>

### 1️⃣ UI is shown to load WordPress and start migration

1. Clean install the WordPress app and authenticate.
2. Exit the WordPress app and don't reopen it. (This will prevent the pre-flight sequence from running)
3. Clean install the Jetpack app and load it.
4. **Expect:** UI to be shown explaining to the user they can open WordPress
5. Tap "Open WordPress"
6. **Expect:** WordPress to load into the foreground temporarily, and then the Jetpack app is brought back to the foreground.
7. **Expect:** The migration view to be shown in Jetpack.

### 2️⃣ User can skip the screen to load WordPress and start migration

1. Clean install the WordPress app and authenticate.
2. Exit the WordPress app and don't reopen it. (This will prevent the pre-flight sequence from running)
3. Clean install the Jetpack app and load it.
4. **Expect:** UI to be shown explaining to the user they can open WordPress
5. Tap "No thanks"
6. **Expect:** The Jetpack landing screen to be shown.

## Regression Notes
1. Potential unintended areas of impact
This flow should only apply to users with a compatible WordPress app but never open it once. Any other scenario is a bug.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing

3. What automated tests I added (or what prevented me from doing so)
Added tests for changes in the `DataMigrator`

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

dvdchr committed:
Add flag after data export succeeded

dvdchr committed:
Add a new dataNotReadyForImport error

dvdchr committed:
Refactor JetpackWindowManager, provide method that handles migration error

dvdchr committed:
Add route for WP that triggers the export process and bounce back to JP

dvdchr committed:
Add tests for DataMigrator

twstokes committed:
Merge remote-tracking branch 'origin/release/21.3' into feature/cmf-remove-export-flag-after-import

twstokes committed:
Merge branch 'feature/cmf-remove-export-flag-after-import' of https://github.com/wordpress-mobile/WordPress-iOS into feature/cmf-remove-export-flag-after-import

twstokes commented on WordPress/Jetpack/Classes/System/JetpackWindowManager.swift:
```diff
@@ -43,20 +37,35 @@ class JetpackWindowManager: WindowManager {
+    /// Checks whether the WordPress app supports the custom scheme meant to disable notifications.
+    /// Since the scheme is added in 21.3, we can be pretty sure that the WP version also supports migration.
+    var isCompatibleWordPressAppPresent: Bool {
+        JetpackNotificationMigrationService.shared.isMigrationSupported
```
I'd thought about using that notifications scheme for this detection [when implementing this](https://github.com/wordpress-mobile/WordPress-iOS/blob/5ab373980120459dd3e3bc5e166560f50512f6a3/WordPress/Jetpack/Classes/ViewRelated/WordPress-to-Jetpack%20Migration/Common/MigrationAppDetection.swift#L17), but didn't know how tightly coupled they were and if we'd need to make a "V2" of the migration procedure down the road.

dvdchr replied:
Oh, good point. I'll use the `MigrationAppDetection`instead!

twstokes committed:
Inject state into the WP migratable state tracker.

twstokes committed:
Show the MigrationLoadWordPress view controller when we detect that WordPress is migratible but needs to run pre-flight.

twstokes commented on WordPress/Jetpack/Classes/System/JetpackWindowManager.swift:
```diff
@@ -73,12 +82,43 @@ class JetpackWindowManager: WindowManager {
+            showSignInUI()
+            return
+        }
+
```
I'll break this out into a function to clean things up.

dvdchr replied:
Sounds good!

dvdchr committed:
Switch to wordPressMigrationV1 scheme

dvdchr commented on WordPress/Classes/System/WordPressAppDelegate+openURL.swift:
```diff
@@ -20,6 +20,19 @@ import AutomatticTracks
+        /// handle the migration-related deeplink intents.
+        /// `UniversalLinkRouter` requires that it's in a universal link format (https://wordpress.com/<path>),
+        /// so we'll hackily converts the deeplink into an acceptable form.
+        if url.scheme == AppScheme.wordpressMigrationV1.rawValue,
```
@twstokes I'll need to add the `UIApplication+Availability` here to check the scheme before executing the export on the WP side. Do you have any concerns about making `AppScheme` available to WP?

twstokes replied:
Not at all! Go for it. 👍

dvdchr replied:
Done ✅

dvdchr commented on WordPress/Classes/Utility/Universal Links/Routes+Jetpack.swift:
```diff
@@ -12,3 +12,35 @@ extension JetpackRoute: NavigationAction {
+            // Regardless of the result, redirect the user back to Jetpack.
+            let jetpackUrl: URL? = {
+                var components = URLComponents()
+                components.scheme = JetpackNotificationMigrationService.jetpackScheme
```
I should probably change this to something that's less dependant on the notification scheme.

dvdchr replied:
On second thought, changing this to `jetpack://` might inadvertently open the game instead 😆 . So I'll just keep this as is for now.

dvdchr commented on WordPress/Jetpack/Classes/System/JetpackWindowManager.swift:
```diff
@@ -73,12 +80,43 @@ class JetpackWindowManager: WindowManager {
+            }
+        }
+        self.show(loadWordPressViewController)
+        return
```
I suppose we don't need this `return`?

twstokes replied:
Nope! It was wiped with the method extraction. 🙇

twstokes committed:
Move WordPress loader UI to a separate function and fire analytics earlier.

twstokes committed:
Merge branch 'feature/cmf-remove-export-flag-after-import' of https://github.com/wordpress-mobile/WordPress-iOS into feature/cmf-remove-export-flag-after-import

twstokes commented on WordPress/Jetpack/Classes/System/JetpackWindowManager.swift:
```diff
@@ -17,21 +17,19 @@ class JetpackWindowManager: WindowManager {
-    func importAndShowMigrationContent(_ blog: Blog?, failureCompletion: (() -> ())?) {
+    func importAndShowMigrationContent(_ blog: Blog? = nil) {
+        /// Track if the WordPress installation was a migratable version
+        let wpInstallationState = MigrationAppDetection.getWordPressInstallationState()
```
I moved this analytics call higher so that every time we're attempting to migrate data we track if the WordPress version was compatible. If you know of a better spot just let me know. 👍

dvdchr replied:
![](https://media.giphy.com/media/U2O50cAkpmTjG/giphy.gif)

dvdchr committed:
Properly detect deeplink scheme

twstokes committed:
Update copy from designs.

dvdchr committed:
Add deeplink handling from the JP side, tidy up code

dvdchr committed:
Fix typo

dvdchr committed:
Add routes and the deeplink router to Jetpack target

twstokes commented on WordPress/Classes/System/WordPressAppDelegate+openURL.swift:
```diff
@@ -20,6 +20,22 @@ import AutomatticTracks
             return true
         }
 
+        /// WordPress only. Handle deeplink from JP that requests data export.
```
Would it be helpful to wrap these in a `AppConfiguration.isWordPress` call? I can add it if you think that's a good idea.

dvdchr replied:
Good idea. It's guarded in the route's `perform` method, but if it bounces there would be a potential logic error there because the next line after `handle` returns true.

This is addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/19710/commits/af22570931c20bf9951f44ec5f0d1cb9716d21dc

dvdchr committed:
Add paths to the deeplink URLs triggered in WP and JP

twstokes committed:
Use secondary text for view.

twstokes committed:
Offset header.

dvdchr committed:
Check current app context before processing migration deeplinks

twstokes committed:
Merge remote-tracking branch 'origin/release/21.3' into feature/cmf-remove-export-flag-after-import

twstokes committed:
Merge branch 'feature/cmf-remove-export-flag-after-import' of https://github.com/wordpress-mobile/WordPress-iOS into feature/cmf-remove-export-flag-after-import

salimbraksa commented on WordPress/Jetpack/Classes/System/JetpackWindowManager.swift:
```diff
@@ -21,6 +21,7 @@ class JetpackWindowManager: WindowManager {
 
         guard AccountHelper.isLoggedIn else {
             shouldImportMigrationData ? importAndShowMigrationContent(blog) : showSignInUI()
+            self.migrationTacker.trackContentImportEligibility(eligible: shouldImportMigrationData)
```
I think tracking import eligibility should be called before importing and showing the migration UI.

twstokes replied:
Yep, good catch. Fixed in 2d49153fbda47ec182767cc6f165447b60183aed.

twstokes committed:
Track import eligibility before showing the UI.

twstokes committed:
Correct typo.

salimbraksa commented on WordPress/Jetpack/Classes/ViewRelated/WordPress-to-Jetpack Migration/Load WordPress/MigrationLoadWordPressViewController.swift:
```diff
@@ -0,0 +1,36 @@
+            actionsView: MigrationActionsView(configuration: viewModel.actions),
+            centerView: UIView()
+        )
+        migrationView.additionalContentInset.top = 30
```
@dvdchr @twstokes If this view controller doesn't have a navigation bar, I think we can remove this line to keep the UI consistent with other screens that don't show the navigation bar. The default value of `additionalContentInset.top` is `UINavigationBar().intrinsicContentSize.height` ( 44pt ).

twstokes replied:
Cool! Completed in 4efcab0d910da913ffcec8fc60c99cb80ae73522.

twstokes committed:
Use default value for offset.

dvdchr committed:
Remove JetpackImportRoute

Jetpack will check and perform import on every willForeground call, which should cover this route's use case.

salimbraksa commented on WordPress/Jetpack/Classes/ViewRelated/WordPress-to-Jetpack Migration/Common/MigratableStateTracker.swift:
```diff
@@ -14,14 +14,14 @@ struct MigratableStateTracker {
+            /// Not tracked.
             break
         }
     }
```
I think this file is not needed anymore, right?

twstokes replied:
Yep, this was a remnant of the release branch merge. Wiped in 566646717bb9f93d1a0f2419f7c20380afbc96ab.

twstokes committed:
Remove old state tracker.

twstokes committed:
Merge branch 'feature/cmf-remove-export-flag-after-import' of https://github.com/wordpress-mobile/WordPress-iOS into feature/cmf-remove-export-flag-after-import

twstokes committed:
Don't dismiss the VC when loading WP.

salimbraksa commented on WordPress/Jetpack/Classes/System/JetpackWindowManager.swift:
```diff
@@ -39,34 +30,49 @@ class JetpackWindowManager: WindowManager {
-                self.migrationTacker.trackContentImportFailed(reason: error.localizedDescription)
-                failureCompletion?()
+                self.handleMigrationFailure(error)
+                self.migrationTracker.trackContentImportFailed(reason: error.localizedDescription)
```
Same as earlier, would be great to track the failure event before the app navigates to another screen as the order of events may be incorrect.

twstokes replied:
Completed in 524c3716df30aa4249daef40a69ff6a5c8ed0b47. Thanks!

twstokes committed:
Merge remote-tracking branch 'origin/release/21.3' into feature/cmf-remove-export-flag-after-import

twstokes committed:
Tweak ordering of calls.

salimbraksa commented:
Great work @twstokes @dvdchr. I tested both cases and they worked as expected. 👏 

I tested the following edge cases and I'll let you decide whether to address them or not:

1. **WordPress Signed Out**: I signed out from WordPress, I launched the Jetpack app and it showed the "Load WordPress" screen. I tapped "Open WordPress", I was redirected to WordPress app then redirected app to Jetpack app with "Load WordPress" as the visible screen. **I think this edge case is low priority as it's very unlikely the user would be in this situation.**
2. **Simulate Content Export Failure**: I disconnected my internet, then added an unsynced local draft to simulate content export failure. I launched the Jetpack app and it showed the "Load WordPress" screen. I tapped "Open WordPress", I was redirected to WordPress app then redirected app to Jetpack app with "Load WordPress" as the visible screen. In this case, the user may not know why the migration didn't work? Perhaps this edge case should be looked into.

| WordPress Signed Out | Simulate Content Export Failure ( Unsynced Local Drafts ) |
| ---------------------- | ---------------------------------------------------------- |
| <video src="https://user-images.githubusercontent.com/9609223/205391077-3695afdb-d94b-4811-b104-74c1a8f96be4.mp4" /> | <video src="https://user-images.githubusercontent.com/9609223/205391453-0ee82217-a396-4493-a23b-edc418c4cba5.mp4" />

salimbraksa commented:
Regarding the 2nd edge case, it wasn't addressed previously because the "Try the new Jetpack" sheet doesn't mention anything about the migration, so when the user taps the button and the migration flow shows up on Jetpack, it's like an unexpected surprise.

However, the "Load WordPress" does tell the user about the migration, so the user now is expecting the migration to work.

twstokes commented:
Thanks for pointing these out @salimbraksa!

> **WordPress Signed Out**

I think we can add an issue to enhance this experience but not block this PR. The users who experience this will be ones who:
- Have WordPress installed on their device and updated to the latest version
- Have Jetpack installed on their device and updated to the latest version
- Aren't authenticated into either

Which should be a relatively small set.

> **Simulate Content Export Failure**

How do feel about us adding an issue for this as well to determine if it's a blocker for the release?

Thanks! I'll be glad to add issues for both of these.

salimbraksa reviewed (approved):
Sounds good to me. Those edges cases are definitely not blocking 👍

twstokes commented:
WordPress signed out issue: https://github.com/wordpress-mobile/WordPress-iOS/issues/19715
Content export failure issue: https://github.com/wordpress-mobile/WordPress-iOS/issues/19716
